connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java

```diff
@@ -405,13 +471,102 @@ protected ConfigDef config(Predicate<?> predicate) {
+
+        Note: Connect when using service loading has an issue outlined in KAFKA-18119. The issue means that the above
+        logic does not hold currently for clusters using service loading when converters are defined in the connector.
+        However, the logic to determine the default should ideally follow the one outlined above, and the code here
+        should still show the correct default version regardless of the bug.
+        */
+        final String connectorConverter = connectorProps.get(connectorConverterConfig);
+        final String workerConverter = workerProps.get(workerConverterConfig);
+        final String connectorClass = connectorProps.get(ConnectorConfig.CONNECTOR_CLASS_CONFIG);
+        final String connectorVersion = connectorProps.get(ConnectorConfig.CONNECTOR_VERSION);
+        if (connectorClass == null || (connectorConverter == null && workerConverter == null)) {
+            return new ConverterDefaults();
+        }
+
+        ConverterDefaults defaults = new ConverterDefaults();
+        // update the default of connector converter based on if the worker converter is provided
+        defaults.type = workerConverter;
+
+
+        String version = null;
+        if (connectorConverter != null) {
+            version = fetchPluginVersion(plugins, connectorClass, connectorVersion, connectorConverter, converterType);
+        } else {
+            version = workerProps.get(workerConverterVersionConfig);
+            if (version == null) {
+                version = plugins.latestVersion(workerConverter);
+            }
+        }
+        defaults.version = version;
+        return defaults;
+    }
+
+    private static void updateKeyDefault(ConfigDef configDef, String versionConfigKey, String versionDefault) {
+        ConfigDef.ConfigKey key = configDef.configKeys().get(versionConfigKey);
+        if (key == null) {
+            return;
+        }
+        configDef.configKeys().put(versionConfigKey, new ConfigDef.ConfigKey(
+                versionConfigKey, key.type, versionDefault, key.validator, key.importance, key.documentation, key.group, key.orderInGroup, key.width, key.displayName, key.dependents, key.recommender, false
+        ));
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> String fetchPluginVersion(Plugins plugins, String connectorClass, String connectorVersion, String pluginName, Class<T> pluginClass) {
+        if (pluginName == null) {
+            return null;
+        }
+        try {
+            VersionRange range = PluginUtils.connectorVersionRequirement(connectorVersion);
+            ClassLoader connectorLoader = plugins.pluginLoader(connectorClass, range);
+            try (LoaderSwap loaderSwap = plugins.withClassLoader(connectorLoader)) {
+                T plugin = (T) plugins.newPlugin(pluginName, pluginClass, null);
+                if (plugin instanceof Versioned) {
+                    return ((Versioned) plugin).version();
+                }
+            }
```

Comment:
This can be a first-class operation in Plugins (and/or DelegatingClassLoader) because it can be computed from the information collected during scanning, without re-instantiating the plugin.

Like `Plugins#latestVersion(String, VersionRange)`

Reply:
I think `plugin.latestVersion` can be incorrect under some edge cases. If two different version of the same plugin is bundled under the same plugin classloader, latest version will return the higher version. However, the actual classloading will use whichever one it finds earlier.  Although this is not the recommended way to install plugins, we do not check for this and don't fail during plugin scanning. I feel that the only real way of getting the version is to actually instantiate the plugin and get its version.

Reply:
> If two different version of the same plugin is bundled under the same plugin classloader

This is impossible from Connect's perspective. Once a ClassLoader returns a Class, it is cached and cannot be redefined. If two versions of a plugin are both present within the same ClassLoader, one will be picked arbitrarily, and it will always be picked (on the local machine). Effectively, Connect cant even know the second class file is there, which is why we don't try and detect it.

Reply:
Ohh interesting, does this apply to classpath plugins as well i.e. only one of the plugin is loded even if there are multiple? Or, is the difference because we provide just one URL to the pluginloader and JAVA just finds the first class there, whereas for the classpath loading we provide all the URL for each path there? I was thinking that we might have to add the verification logic for versioned plugin that we did for classpath loaded plugins, but this simplifies things.

Reply:
> does this apply to classpath plugins as well i.e. only one of the plugin is loded even if there are multiple?

Yes it applies to the classpath too.